README.md
# pacman4console.debian

This repository contains the **Debian packaging files** for the game
pacman4console (_console based pacman game_).

If you're after the game pacman4console itself, [go to it's homepage][home]

![screen](https://sites.google.com/site/doctormike/pacman-1-1.png)

## Debian Packaging

### Structure

* [**pacman4console on Debian Package Tracking System**][package]

This package is maintained on a Git repository through some important branches.

| branch                                                                            | role |
| --------------------------------------------------------------------------------- | ---- |
| [**debian**](https://github.com/alexdantas/pacman4console.debian/tree/debian)            | pacman4console _and_ the debian package source code (`debian` directory) |
| [**upstream**](https://github.com/alexdantas/pacman4console.debian/tree/upstream)        | Everything that was on the original upstream tarball (`orig.tar.gz`) |
| [**pristine-tar**](https://github.com/alexdantas/pacman4console.debian/tree/pristine-tar)| Debian original tarball |
| [**readme**](https://github.com/alexdantas/pacman4console.debian/tree/readme)            | This single file, explaining all that |


### Getting Started

To start working on this package, clone this repository
 [along with all branches](http://stackoverflow.com/a/72156):

```bash
# Cloning the entire repository
$ git clone https://github.com/alexdantas/pacman4console.debian

# For now you have only the `readme` branch
$ git branch
* readme

# But the other branches are actually "hidden"
$ git branch -a
* readme
remotes/origin/debian
remotes/origin/pristine-tar
remotes/origin/upstream

# So start working on one of them with this
# specific command:
$ git checkout -b debian origin/debian

# And by this moment you have this branch as "non-hidden"
# You can go back and forth as you like
$ git checkout -b readme
$ git checkout -b debian
```

### Notes

1. Remember that the **debian** is _not_ the one you get
   right after cloning this repository!
2. [Workaround.org has a nice article][article] on co-maintaining
a Debian package with Git and `git-buildpackage`.

[home]:    https://sites.google.com/site/doctormike/pacman.html
[package]: https://tracker.debian.org/pkg/pacman4console
[article]: https://workaround.org/debian-git-comaintenance


debian/changelog
pacman4console (1.3-2) UNRELEASED; urgency=medium

  * Use secure copyright file specification URI.
  * Update watch file format version to 4.
  * Bump debhelper from deprecated 9 to 13.
  * Set debhelper-compat version in Build-Depends.
  * Use secure URI in Vcs control header Vcs-Git.
  * Drop unnecessary dh arguments: --parallel

 -- Debian Janitor <janitor@jelmer.uk>  Wed, 02 Nov 2022 13:07:40 -0000

pacman4console (1.3-1) unstable; urgency=medium

  * New upstream release
  * New maintainer (closes: #756024)
  * debian/patches: Updated all patches to newest source code
  * debian/control:
      - Changed Vcs-* fields to new repositories
      - Changed package description

 -- Alexandre Dantas <eu@alexdantas.net>  Tue, 12 Aug 2014 13:32:31 -0400

pacman4console (1.2-8) unstable; urgency=medium

  * Orphaned (see #756024).
  * Changed maintainer to Debian QA Group <packages@qa.debian.org>.

 -- Joao Eriberto Mota Filho <eriberto@eriberto.pro.br>  Fri, 25 Jul 2014 11:24:01 -0300

pacman4console (1.2-7) unstable; urgency=medium

  * debian/desktop/pacman4console: renamed to pacman4console.desktop.
    Thanks to Markus Koschany <apo@gambaru.de>. (Closes: #738709)
  * debian/docs: removed because the screenshot.png file is useless
    and the upstream README is a bit deprecated.
  * debian/install: updated because a file was renamed and another
    was moved to debian/desktop/.
  * debian/man/:
      - Removed the useless file pacman4console.6.old.
      - Updated and improved the pacman4consoleeditor.1.
  * debian/pacman4console.xpm: moved to debian/desktop directory.
  * debian/patches/Makefile: updated to avoid the upstream
    Levels/README install, because the pacman4consoleeditor.1 now has
    this information.
  * debian/README.Debian: added to gather essential information from
    upstream README.
  * debian/rules: removed useless lines 'DPKG_EXPORT_BUILDFLAGS' and
    'include /usr/share/dpkg/buildflags.mk'. Using DH9 turns these
    lines obsoletes.

 -- Joao Eriberto Mota Filho <eriberto@eriberto.pro.br>  Wed, 12 Feb 2014 09:57:18 -0200

pacman4console (1.2-6) unstable; urgency=medium

  * Added a desktop entry. Thanks to Markus Koschany <apo@gambaru.de>
    for the suggestion. So, created debian/{desktop/pacman4cosole,menu,
    pacman4console.xpm}. (Closes: #737997)
  * manpages:
      - debian/man/: created to group all manpages and their sources.
      - pacman4console.1: moved to debian/man/ and updated.
      - pacman4consoleedit.1: moved to debian/man/.
  * debian/copyright: updated the packaging copyright years.
  * debian/install: added to install desktop and pacman4console.xpm files.
  * debian/patches:
      - Added a new patch (pacman.c) that allows exit or restart after
        "game over" message. This was fundamental to add a desktop entry.
        This patch gather the old 'fix_array_index' and 'pacman.c' patches.
        This patch also provides new features to pacman4console. To see more
        details, read the debian/patches/pacman.c from Debian source code.
        Special thanks to Yannic Scheper <ys42@cd42.de> that replied me
        quickly and made all necessary changes, creating a patch in some
        minutes, once the upstream stopped the development.
      - fix_array_index: deleted because the pacman.c provides the same
        function.
      - Removed the .patch extension from all patches.
  * debian/README.source: added to explain about new patch.
  * debian/source/lintian-overrides: removed as requested by Paul Wise.
    Thanks to Paul for explanations about the idea of the GPG sign check.
  * debian/watch: improved.

 -- Joao Eriberto Mota Filho <eriberto@eriberto.pro.br>  Sun, 09 Feb 2014 13:11:09 -0200

pacman4console (1.2-5) unstable; urgency=medium

  * Bumped Standards-Version from 3.9.4 to 3.9.5.
  * debian/copyright: added the upstream e-mail address.
  * debian/control:
      - Changed Priority from extra to optional.
      - Fixed the Vcs-Browser field.
  * debian/source/: added an override to reply to check-gpg-signature.
  * debian/watch: improved.

 -- Joao Eriberto Mota Filho <eriberto@eriberto.pro.br>  Fri, 27 Dec 2013 14:53:25 -0200

pacman4console (1.2-4) unstable; urgency=low

  * debian/watch: re-added. Thanks a lot to Bart Martens <bartm@debian.org>.
  * debian/source/lintian-overrides: removed because it is useless now.

 -- Joao Eriberto Mota Filho <eriberto@eriberto.pro.br>  Thu, 25 Jul 2013 11:42:24 -0300

pacman4console (1.2-3) unstable; urgency=low

  * Migrations:
      - debian format from 1.0 to 3.0. (Closes: #661027)
      - debian/changelog to 1.0.
      - debhelper level from 5 to 9.
      - patch system from dpatch to quilt.
      - VCS: svn to git.
  * debian/control:
      - added dpkg-dev as Build-Depends.
      - added ${misc:Depends} to Depends field.
      - droped leading article from short description.
      - removed screeshot reference from long description.
      - updated Standards-Version from 3.7.3 to 3.9.4.
      - updated homepage.
  * debian/dirs: removed. Using Makefile only.
  * debian/gbp.conf: added to allow git-buildpackage use.
  * debian/manpages: created to install manuals.
  * debian/override: renamed to lintian-overrides.
  * debian/patches/
      - fix_array_index.patch: Fix array index out of bounds if pacman or
            ghost is going through one border to the opposite site. Thanks
            to Yannic Scheper. (Closes: #641657)
      - Makefile.patch:
          . added GCC hardening.
          . adjusted lncurses position to avoid FTBFS when using --as-needed.
            Thanks to Alex Chiang. (Closes: #641384)
          . switched several commands by "install" to simplify actions.
      - pacman.c.patch: fix possible buffer overflow when using an argument
            larger than 99 characters to start the program. Thanks to Yannic
            Scheper. (Closes: #641652)
  * debian/rules: updated to new (reduced) format.

pacmanedit.c
/****************************************************
* Pacman For Console Level Editor                   *
* By: Mike Billars (michaelbillars@gmail.com)       *
* Date: 2014-04-26                                  *
*                                                   *
* Please see file COPYING for details on licensing  *
*       and redistribution of this program          *
*                                                   *
****************************************************/

/***********
* INCLUDES *
***********/
#include <stdio.h>
#include <curses.h>
#include <string.h>
#include <stdlib.h>

/*************
* PROTOTYPES *
*************/
void CheckScreenSize();                                 //Make sure resolution is at least 32x29
void CreateWindows(int y, int x, int y0, int x0);       //Make ncurses windows
void DrawWindow();                                      //Refresh display
void ExitProgram(char message[255]);                    //Exit gracefully
void GetInput();                                        //Get user input
void InitCurses();                                      //Start up ncurses
void LoadLevel();                                       //Load level into memory
void MainLoop();                                        //Main loop
void SaveLevel();                                       //Save level to specified file

/*******************
* GLOBAL VARIABLES *
*******************/
//I know global variables are bad, but it's just sooo easy!

//Windows used by ncurses
WINDOW * win;
WINDOW * status;

//For colors
enum { Wall = 1, Normal, Pellet, PowerUp, GhostWall, Ghost1, Ghost2, Ghost3, Ghost4, BlueGhost, Pacman, Cursor};

int Level[29][28] = { 0 };                              //Main level array
int Loc[6][2] = { 0 };                                  //Ghosts, Pacman, and cursor locations
int tleft = 0;                                          //How long left for invincibility
char filename[255] = "";                                //Name of file to load/save

/****************************************************************
* Function:    main()                                           *
* Parameters:  argc, argv (passed from command line)            *
* Returns:     Success                                          *
* Description: Initiate the program and call the subfunctions   *
****************************************************************/
int main(int argc, char *argv[100]) {

    //Make sure they entered a filename to open/save
    if((argc > 1) && (strlen(argv[1]) >= 1)) {

        strcpy(filename, argv[1]);                      //Copy command line parameter into string
        LoadLevel();                                    //Load the file

        InitCurses();                                   //Initialize ncurses
        CheckScreenSize();                              //Make sure the screen is big enough
        CreateWindows(29, 28, 1, 1);                    //

        DrawWindow();
        Loc[5][0]=13;
        Loc[5][1]=13;
        MainLoop();
    }
    else
        ExitProgram("You must specify a file to use (load and save)");

}

/****************************************************************
* Function:    CheckScreenSize()                                *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Make sure the virtual console is big enough      *
****************************************************************/
void CheckScreenSize() {

    //Make sure the window is big enough
    int h, w; getmaxyx(stdscr, h, w);

    if((h < 32) || (w < 29)) {
        endwin();
        fprintf(stderr, "\nSorry.\n");
        fprintf(stderr, "To edit levels for Pacman for Console, your console window\n");
        fprintf(stderr, "must be at least 32x29 Please resize your window/resolution\n");
        fprintf(stderr, "and re-run the program.\n\n");
        ExitProgram("");
    }
}

/****************************************************************
* Function:    CreateWindows()                                  *
* Parameters:  y, x, y0, x0 (coords and size of window)         *
* Returns:     none                                             *
* Description: Create the main game windows                     *
****************************************************************/
void CreateWindows(int y, int x, int y0, int x0) {
    //Create two new windows, for status bar and for main level
    win = newwin(y, x, y0, x0);
    status = newwin(3, 28, 29, 1);
}

/****************************************************************
* Function:    DrawWindow()                                     *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Redraw each window to update the screen          *
****************************************************************/
void DrawWindow() {
    int a = 0; int b = 0;
    char chr = ' ';
    int attr;

    //Display level array
    for(a = 0; a < 29; a++) for(b = 0; b < 28; b++) {
        switch(Level[a][b]) {
        case 0: chr = ' '; attr = A_NORMAL; wattron(win, COLOR_PAIR(Normal));    break;
        case 1: chr = ' '; attr = A_NORMAL; wattron(win, COLOR_PAIR(Wall));      break;
        case 2: chr = '.'; attr = A_NORMAL; wattron(win, COLOR_PAIR(Pellet));    break;
        case 3: chr = '*'; attr = A_BOLD;   wattron(win, COLOR_PAIR(PowerUp));   break;
        case 4: chr = ' '; attr = A_NORMAL; wattron(win, COLOR_PAIR(GhostWall)); break;
        case 5: chr = '&'; attr = A_NORMAL; wattron(win, COLOR_PAIR(Ghost1));    break;
        case 6: chr = '&'; attr = A_NORMAL; wattron(win, COLOR_PAIR(Ghost2));    break;
        case 7: chr = '&'; attr = A_NORMAL; wattron(win, COLOR_PAIR(Ghost3));    break;
        case 8: chr = '&'; attr = A_NORMAL; wattron(win, COLOR_PAIR(Ghost4));    break;
        case 9: chr = 'C'; attr = A_NORMAL; wattron(win, COLOR_PAIR(Pacman));    break;
        }
        mvwaddch(win, a, b, chr | attr);
    }

    //Display the cursor
    wattron(win,COLOR_PAIR(Cursor));
    mvwaddch(win, Loc[5][0], Loc[5][1], 'X');

    //Display info in status
    wmove(status, 1, 1);
    wattron(status, COLOR_PAIR(Normal));
    mvwprintw(status, 1,  0, "0)     1)       2)     3)");
    mvwprintw(status, 2,  0, "4)     5-8)     9)     F/C/Q");

    wattron(status, COLOR_PAIR(Normal));    mvwaddch(status, 1,  3, ' ');
    wattron(status, COLOR_PAIR(Wall));      mvwaddch(status, 1, 10, ' ');
    wattron(status, COLOR_PAIR(Pellet));    mvwaddch(status, 1, 19, '.');
    wattron(status, COLOR_PAIR(PowerUp));   mvwaddch(status, 1, 26 , '*' | A_BOLD);
    wattron(status, COLOR_PAIR(GhostWall)); mvwaddch(status, 2,  3, ' ');
    wattron(status, COLOR_PAIR(Ghost1));    mvwaddch(status, 2, 12, '&');
    wattron(status, COLOR_PAIR(Pacman));    mvwaddch(status, 2, 19, 'C');

    wrefresh(status);
    wrefresh(win);
}

/****************************************************************
* Function:    ExitProgram()                                    *
* Parameters:  message: text to display before exiting          *
* Returns:     none                                             *
* Description: Gracefully end program                           *
****************************************************************/
void ExitProgram(char message[255]) {

    endwin();

    //Must save file
    if(message[0]=='s') {
        printf("\n\nFile saved as: %s\n", filename);
    }

    else
        printf("%s\n\n", message);

    exit(0);
}

/****************************************************************
* Function:    GetInput()                                       *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Get input from user and take appropriate action  *
****************************************************************/
void GetInput() {
    int ch;
    int a=0;
    int b=0;

    static int chtmp;

    ch = getch();

    //Buffer input
    do {
        DrawWindow();
        ch = getch();
    } while (ch == ERR);

    switch (ch) {
    case KEY_UP:    case 'w': case 'W':
        Loc[5][0]--;
        if(Loc[5][0]<0) { Loc[5][0]=28; }
        break;

    case KEY_DOWN:  case 's': case 'S':
        Loc[5][0]++;
        if(Loc[5][0]>28) { Loc[5][0]=0; }
        break;

    case KEY_LEFT:  case 'a': case 'A':
        Loc[5][1]--;
        if(Loc[5][1]<0) { Loc[5][1]=27; }
        break;

    case KEY_RIGHT: case 'd': case 'D':
        Loc[5][1]++;
        if(Loc[5][1]>27) { Loc[5][1]=0; }
        break;

    case 'q': case 'Q':
        //Save and exit
        SaveLevel();
        ExitProgram("s");
        break;

    //Draw a blank space
    case '0': case ' ':
        Level[Loc[5][0]][Loc[5][1]]=0;
        break;

    //Draw a wall
    case '1':
        Level[Loc[5][0]][Loc[5][1]]=1;
        break;

    //Draw a pellet
    case '2':
        Level[Loc[5][0]][Loc[5][1]]=2;
        break;

    //Draw a powerup
    case '3':
        Level[Loc[5][0]][Loc[5][1]]=3;
        break;

    //Draw a ghost/spawning wall
    case '4':
        Level[Loc[5][0]][Loc[5][1]]=4;
        break;

    //Draw a ghost/pacman
    case '5': case '6': case '7':
    case '8': case '9':
        //Move ghost/pacman to new location
        Level[Loc[5][0]][Loc[5][1]]=ch-'0';

        //Erase character from previous location and replace with BLANK
        if(Level[Loc[ch-'0'-5][0]][Loc[ch-'0'-5][1]]==ch-'0')
            Level[Loc[ch-'0'-5][0]][Loc[ch-'0'-5][1]]=0;

        //Set Loc variable to new location
        Loc[ch-'0'-5][0]=Loc[5][0];
        Loc[ch-'0'-5][1]=Loc[5][1];

        break;

    //Fill blank spaces with pellets
    case 'f': case 'F':
        for(a=0; a<29; a++)
        for(b=0; b<28; b++)
            if(Level[a][b]==0) Level[a][b]=2;
        break;

    //Fill pellet spaces with blanks
    case 'c': case 'C':
        for(a=0; a<29; a++)
        for(b=0; b<28; b++)
            if(Level[a][b]==2) Level[a][b]=0;
        break;

    }
}

/****************************************************************
* Function:    InitCurses()                                     *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Initialize ncurses and set defined colors        *
****************************************************************/
void InitCurses() {

    initscr();                      //Needed for ncurses windows
    start_color();                  //Activate colors in console
    curs_set(0);                    //      Don't remember
    keypad(stdscr, TRUE);           //Activate arrow keys
    nodelay(stdscr, TRUE);          //Allow getch to work without pausing
    nonl();                         //      Don't remember
    cbreak();                       //      Don't remember
    noecho();                       //Don't display input key

    //Make custom text colors
    init_pair(Normal,    COLOR_WHITE,   COLOR_BLACK);
    init_pair(Wall,      COLOR_WHITE,   COLOR_WHITE);
    init_pair(Pellet,    COLOR_WHITE,   COLOR_BLACK);
    init_pair(PowerUp,   COLOR_BLUE,    COLOR_BLACK);
    init_pair(GhostWall, COLOR_WHITE,   COLOR_CYAN);
    init_pair(Ghost1,    COLOR_RED,     COLOR_BLACK);
    init_pair(Ghost2,    COLOR_CYAN,    COLOR_BLACK);
    init_pair(Ghost3,    COLOR_MAGENTA, COLOR_BLACK);
    init_pair(Ghost4,    COLOR_YELLOW,  COLOR_BLACK);
    init_pair(BlueGhost, COLOR_BLUE,    COLOR_RED);
    init_pair(Pacman,    COLOR_YELLOW,  COLOR_BLACK);
    init_pair(Cursor,    COLOR_BLUE,    COLOR_YELLOW);
}

/****************************************************************
* Function:    LoadLevel()                                      *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Open level file and load it into memory          *
****************************************************************/
void LoadLevel() {

    int a = 0; int b = 0;
    FILE *fin;

    //Open file
    fin = fopen(filename, "r");

    //File doesn't exist, start new
    if(!(fin)) {
        for(a=0; a<28; a++) { Level[a][0]=1; Level[a][27]=1; }
        for(a=0; a<28; a++) { Level[0][a]=1; Level[28][a]=1; }
        Loc[0][0]=2; Loc[0][1]=2;
        Loc[1][0]=2; Loc[1][1]=4;
        Loc[2][0]=2; Loc[2][1]=6;
        Loc[3][0]=2; Loc[3][1]=8;
        Loc[4][0]=19; Loc[4][1]=19;
        Level[Loc[0][0]][Loc[0][1]]=5;
        Level[Loc[1][0]][Loc[1][1]]=6;
        Level[Loc[2][0]][Loc[2][1]]=7;
        Level[Loc[3][0]][Loc[3][1]]=8;
        Level[Loc[4][0]][Loc[4][1]]=9;
    }
    else {

        //Open file and load the level into the array
        for(a = 0; a < 29; a++)
        for(b = 0; b < 28; b++) {
            fscanf(fin, "%d", &Level[a][b]);
            if(Level[a][b] == 5) { Loc[0][0]=a; Loc[0][1]=b; }
            if(Level[a][b] == 6) { Loc[1][0]=a; Loc[1][1]=b; }
            if(Level[a][b] == 7) { Loc[2][0]=a; Loc[2][1]=b; }
            if(Level[a][b] == 8) { Loc[3][0]=a; Loc[3][1]=b; }
            if(Level[a][b] == 9) { Loc[4][0]=a; Loc[4][1]=b; }
        }
    }
}

/****************************************************************
* Function:    MainLoop()                                       *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Control the main execution of the game           *
****************************************************************/
void MainLoop() {
    do {
        GetInput();
        DrawWindow();
    } while (1);
}

/****************************************************************
* Function:    SaveLevel()                                      *
* Parameters:  none                                             *
* Returns:     none                                             *
* Description: Save current level array to a specified file     *
****************************************************************/
void SaveLevel() {

    int a=0;
    int b=0;
    FILE *fout;

    //Open file
    fout = fopen(filename, "w");

    //Check for write permissions
    if(!(fout)) {
        ExitProgram("Cannot write file. :-(");
    }

    else {
        for(a = 0; a < 29; a++) {
            for(b = 0; b < 28; b++) {
                fprintf(fout, "%d ", Level[a][b]);
            }
            fprintf(fout, "\n");
        }
        fprintf(fout, "1");
        fclose(fout);
    }
}


debian/patches/pacman.c
Description: fix some problems and add features.
               - Problems fixed:
                   * Changed screen/window resolution error message.
                   * Bugfix for possible buffer overflow when choosing a
                     level. (Closes #641652)
                   * Bugfix for possible index out of bound when going through
                     a border while playing. (Closes #641657)
               - Feature to allow create a menu icon for X (Closes: #737997):
                   * Game Over screen which allows to exit or to restart.
               - Another features:
                   * Added a help option (documents before undocumented level
                     choosing option).
                   * Changed option syntax for level choosing.
Author: Yannic Scheper <ys42@cd42.de>
Last-Update: 2014-08-12
--- pacman4console.orig/pacman.c
+++ pacman4console/pacman.c
@@ -31,7 +31,7 @@
 * PROTOTYPES *
 *************/
 void IntroScreen();                                     //Show introduction screen and menu
-void CheckCollision();                                  //See if Pacman and Ghosts collided
+int  CheckCollision();                                  //See if Pacman and Ghosts collided
 void CheckScreenSize();                                 //Make sure resolution is at least 32x29
 void CreateWindows(int y, int x, int y0, int x0);       //Make ncurses windows
 void Delay();                                           //Slow down game for better control
@@ -40,11 +40,11 @@ void ExitProgram(const char *message);
 void GetInput();                                        //Get user input
 void InitCurses();                                      //Start up ncurses
 void LoadLevel(char *levelfile);                        //Load level into memory
-void MainLoop();                                        //Main program function
+int  MainLoop();                                        //Main program function
 void MoveGhosts();                                      //Update Ghosts' location
 void MovePacman();                                      //Update Pacman's location
 void PauseGame();                                       //Pause
-
+void PrintHelp(char* name);                             //Print help and exit
 
 /*******************
 * GLOBAL VARIABLES *
@@ -76,41 +76,71 @@ int tleft = 0;
 ****************************************************************/
 int main(int argc, char *argv[100]) {
 
-    int j = 0;
+    int loop = 1;       //loop program? 0 = no, 1 = yes
+    char* level = NULL; //level to load
+    int j = 1;
+    int i;
+    for(i = 1; i < argc; ++i) {
+        if(strcmp(argv[i], "--help") == 0 || strcmp(argv[i], "-h") == 0) {
+            PrintHelp(argv[0]);
+            return 0;
+        }
+        else if(strncmp(argv[i], "--level=", 8) == 0) {
+            level = argv[i];
+            level += 8;
+            int len = strlen(level);
+            if(len == 0) {
+                level = NULL;
+            }
+            else if(len == 1) {
+                for(LevelNumber = '1'; LevelNumber <= '9'; LevelNumber++) {
+                    if(LevelNumber == level[0]) {
+                        j = LevelNumber - '0';
+                        level = NULL;
+                        break;
+                    }
+                }
+            }
+        }
+        else {
+            PrintHelp(argv[0]);
+            return 0;
+        }
+    }
+
     srand( (unsigned)time( NULL ) );
 
     InitCurses();                   //Must be called to start ncurses
     CheckScreenSize();              //Make sure screen is big enough
     CreateWindows(29, 28, 1, 1);    //Create the main and status windows
 
-    //If they specified a level to load
-    if((argc > 1) && (strlen(argv[1]) > 1)) {
-        argv[1][99] = '\0';
-        LoadLevel(argv[1]);         //Load it and...
-        MainLoop();                 //Start the game
-    }
-
-    //If they did not enter a level, display intro screen then use default levels
-    else {
-        IntroScreen();              //Show intro "movie"
-        j = 1;                      //Set initial level to 1
-
-        if(argc > 1) j = argv[1][0] - '0';    //They specified a level on which to start (1-9)
-
-        //Load 9 levels, 1 by 1, if you can beat all 9 levels in a row, you're awesome
-        for(LevelNumber = j; LevelNumber < 10; LevelNumber++) {
-
-            //Replace level string underscore with the actual level number (see pacman.h)
-            LevelFile[strlen(LevelFile) - 6] = '0';
-            LevelFile[strlen(LevelFile) - 5] = LevelNumber + '0';
-
-            LoadLevel(LevelFile);   //Load level into memory
-            Invincible = 0;         //Reset invincibility with each new level
-            MainLoop();             //Start the level
+    IntroScreen();              //Show intro "movie"
 
+    int start_lives = Lives;
+    int start_points = Points;
+    do {
+        Lives = start_lives;
+        Points = start_points;
+        if(level == NULL) {
+            //j = 1;
+            //Load levels, 1 by 1, if you can beat all 9 levels in a row, you're awesome
+            for(LevelNumber = j; LevelNumber < 10; LevelNumber++) {
+                            LevelFile[strlen(LevelFile) - 6] = '0';
+                LevelFile[strlen(LevelFile) - 5] = LevelNumber + '0';
+                LoadLevel(LevelFile);
+                Invincible = 0;            //Reset invincibility
+                if(MainLoop() == 1) break;
+            }
         }
+        else {
+            //Load special non-standard level
+            LoadLevel(level);
+             Invincible = 0;            //Reset invincibility
+             MainLoop();
+         }
 
-    }
+        if(GameOverScreen() == 1) loop = 0;
+    } while(loop == 1);
 
     //Game has ended, deactivate and end program
     ExitProgram(EXIT_MSG);
@@ -125,7 +155,7 @@ int main(int argc, char *argv[100]) {
 * Returns:     none                                             *
 * Description: Check and handle if Pacman collided with a ghost *
 ****************************************************************/
-void CheckCollision() {
+int CheckCollision() {
 
     //Temporary variable
     int a = 0;
@@ -165,7 +195,7 @@ void CheckCollision() {
                 usleep(1000000);
 
                 //If no more lives, game over
-                if(Lives == -1) ExitProgram(END_MSG);
+                if(Lives == -1) return 1;
 
                 //If NOT game over...
 
@@ -187,6 +217,7 @@ void CheckCollision() {
             }
         }
     }
+    return 0;
 }
 
 /****************************************************************
@@ -203,13 +234,41 @@ void CheckScreenSize() {
         if((h < 32) || (w < 29)) {
                 endwin();
                 fprintf(stderr, "\nSorry.\n");
-                fprintf(stderr, "To play Pacman for Console, your console window must be at least 32x29\n");
+                fprintf(stderr, "To play Pacman for Console, your console window must be at least 29x32\n");
                 fprintf(stderr, "Please resize your window/resolution and re-run the game.\n\n");
                 exit(0);
         }
 
 }
 
+int GameOverScreen() {
+    char chr = ' ';
+    int a, b;
+    for(a = 0; a < 29; a++) for(b = 0; b < 28; b++) {
+        mvwaddch(win, a, b, chr);
+    }
+
+    wattron(win, COLOR_PAIR(Pacman));
+    mvwprintw(win, 8, 11, "Game Over");
+
+    wattron(win, COLOR_PAIR(Normal));
+    mvwprintw(win, 14, 2, "Press q to quit ...");
+    mvwprintw(win, 16, 2, "... or any other key");
+    mvwprintw(win, 17, 6, "to play again");
+
+    wrefresh(win);
+
+    //And wait
+    int chtmp;
+    do {
+        chtmp = getch();
+    } while (chtmp == ERR);
+
+    if(chtmp == 'q' || chtmp == 'Q')
+        return 1;
+    return 0;
+}
+
 /****************************************************************
 * Function:    CreateWindows()                                  *
 * Parameters:  y, x, y0, x0 (coords and size of window)         *
@@ -494,7 +553,7 @@ void IntroScreen() {
 * Returns:     none                                             *
 * Description: Open level file and load it into memory          *
 ****************************************************************/
-void LoadLevel(char levelfile[100]) {
+void LoadLevel(char* levelfile) {
 
     int a = 0; int b = 0;
     size_t l;
@@ -555,7 +614,7 @@ void LoadLevel(char levelfile[100]) {
 * Returns:     none                                             *
 * Description: Control the main execution of the game           *
 ****************************************************************/
-void MainLoop() {
+int MainLoop() {
 
     DrawWindow();                    //Draw the screen
     wrefresh(win); wrefresh(status); //Refresh it just to make sure
@@ -564,15 +623,15 @@ void MainLoop() {
     /* Move Pacman. Move ghosts. Check for extra life awarded
     from points. Pause for a brief moment. Repeat until all pellets are eaten */
     do {
-        MovePacman();    DrawWindow();    CheckCollision();
-        MoveGhosts();    DrawWindow();    CheckCollision();
+        MovePacman();    DrawWindow();    if (CheckCollision() == 1) return 1;
+        MoveGhosts();    DrawWindow();    if (CheckCollision() == 1) return 1;
         if(Points > FreeLife) { Lives++; FreeLife *= 2;}
         Delay();
     } while (Food > 0);
 
     DrawWindow();                   //Redraw window and...
     usleep(1000000);                //Pause, level complete
-
+    return 0;
 }
 
 /****************************************************************
@@ -748,3 +807,12 @@ void PauseGame() {
     } while (chtmp == ERR);
 
 }
+
+void PrintHelp(char* name) {
+    printf("Usage: %s [OPTION]\n\n", name);
+    printf("Options:\n");
+    printf("  -h, --help        print help\n");
+    printf("  --level=[1..9]    start at specified standard level\n");
+    printf("  --level=LEVEL     play specified non-standard LEVEL\n");
+}
+
